File: other/Solution213.java
class Solution213 {
    public int rob(int[] nums) {
    	if(nums.length == 1){
    		return nums[0];
    	}
    	if(nums.length == 2){
    		return Math.max(nums[0], nums[1]);
    	}
    	int[] dp1 = new int[nums.length - 1];
    	int[] dp2 = new int[nums.length - 1];
    	dp1[0] = nums[0];
    	dp1[1] = Math.max(nums[0], nums[1]);
    	for(int i = 2; i < nums.length - 1; i++){
    		dp1[i] = Math.max(dp1[i - 1], dp1[i - 2] + nums[i]);
    	}
    	dp2[0] = nums[1];
    	dp2[1] = Math.max(nums[1], nums[2]);
    	for(int i = 3; i < nums.length; i++){
    		dp2[i - 1] = Math.max(dp2[i - 2], dp2[i - 3] + nums[i]);
    	}
    	return Math.max(dp1[dp1.length - 1], dp2[dp2.length - 1]);
    }
}

File: C++/Solution213.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution213 {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];
        vector<int> copy1(nums.begin(), nums.end() - 1), copy2(nums.begin() + 1, nums.end());
        return max(rob2(copy1), rob2(copy2));
    }

    int rob2(vector<int> &nums){
        int size = nums.size();
        vector<int> dp(size);
        dp[0] = nums[0];
        if (size == 1) return nums[0];
        dp[1] = max(nums[0], nums[1]);
        for (int i = 2; i < size; i++){
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i]);
        }
        return dp[size - 1];
    }
};